fastfield_codecs/src/compact_space.rs

```diff
@@ -0,0 +1,683 @@
+    ///
+    /// and we want a mapping for 1005, there is no equivalent compact space. We instead return an
+    /// error with the index of the next range.
+    fn to_compact(&self, value: u128) -> Result<u64, usize> {
+        self.params.compact_space.to_compact(value)
+    }
+
+    fn compact_to_u128(&self, compact: u64) -> u128 {
+        self.params.compact_space.unpack(compact)
+    }
+
+    /// Comparing on compact space: 1.2 GElements/s
+    ///
+    /// Comparing on original space: .06 GElements/s (not completely optimized)
+    pub fn get_range(&self, mut range: RangeInclusive<u128>) -> Vec<u64> {
+        if range.start() > range.end() {
+            range = *range.end()..=*range.start();
+        }
+        let from_value = *range.start();
+        let to_value = *range.end();
+        assert!(to_value >= from_value);
+        let compact_from = self.to_compact(from_value);
+        let compact_to = self.to_compact(to_value);
+
+        // Quick return, if both ranges fall into the same non-mapped space, the range can't cover
+        // any values, so we can early exit
+        match (compact_to, compact_from) {
+            (Err(pos1), Err(pos2)) if pos1 == pos2 => return Vec::new(),
+            _ => {}
+        }
+
+        let compact_from = compact_from.unwrap_or_else(|pos| {
+            // Correctness: Out of bounds, if this value is Err(last_index + 1), we early exit,
+            // since the to_value also mapps into the same non-mapped space
+            let range_and_compact_start =
+                self.params.compact_space.get_range_and_compact_start(pos);
+            range_and_compact_start.1
+        });
+        // If there is no compact space, we go to the closest upperbound compact space
+        let compact_to = compact_to.unwrap_or_else(|pos| {
+            // Correctness: Overflow, if this value is Err(0), we early exit,
+            // since the from_value also mapps into the same non-mapped space
+
+            // get end of previous range
+            let pos = pos - 1;
+            let range_and_compact_start =
+                self.params.compact_space.get_range_and_compact_start(pos);
+            let compact_end = range_and_compact_start.1
+                + (range_and_compact_start.0.end() - range_and_compact_start.0.start()) as u64;
+            compact_end
+        });
+
+        let range = compact_from..=compact_to;
+        let mut positions = Vec::new();
+
+        for (pos, compact_value) in self
+            .iter_compact()
+            .enumerate()
+            .filter(|(_pos, val)| *val != self.params.null_value_compact_space)
+        {
+            if range.contains(&compact_value) {
+                positions.push(pos as u64);
+            }
+        }
+
+        positions
+    }
+
+    #[inline]
+    fn iter_compact(&self) -> impl Iterator<Item = u64> + '_ {
+        (0..self.params.num_vals)
+            .map(move |idx| self.params.bit_unpacker.get(idx as u64, &self.data) as u64)
```

Comment:
we can probably optimize this.

Reply:
loop unrolling or do have something else in mind?

Reply:
I did some unrolling, it's 3 times faster. 

Currently the compact space is u64. We could switch to u32. It's unlikely that we hit the u32 boundary, but it can happen (around 1 billion elements in a segment with worse case number spreading).

To apply some simd, we would need to copy the data blockwise into an array. I suspect it will not help though. I never tried to intersperse simd with normal code.

In the bitpacker there is a check which we should get rid of in the loop
```
        if self.num_bits == 0 {
            return 0u64;
        }
```

Reply:
The bitunpacking code is inefficient, and the filtering is inefficient.
I thought we could do better by fill a buffer of size 128 or something, and use simd to postfilter....
but 1Gel/s is already quite nice.